Issue #89: Update documentation for detect-secrets (#95)
NASA-AMMOS/slim: continuous-testing/starter-kits/README.md

This page contains starter kit information, which represent templates, code and configuration to help you get started quickly with continuous testing best practices described in this overall guide. Please see categories and links below for details. 

## Static Application Security Testing

This section contains links to sample actions, templates and configurations that analyze and validate code for security flaws and sensitive information. Identifying security vulnerabilities and sensitive data is an [OSS cybersecurity](https://www.cisa.gov/uscert/ncas/alerts/aa22-137a) [best practice](https://appel.nasa.gov/2022/06/30/spotlight-on-lessons-learned-open-source-and-commercial-web-software-vulnerabilities/).  

### Detect Secrets 

> [detect-secrets](https://github.com/Yelp/detect-secrets), an open-source tool employed and recommended by leading technology companies such as [Microsoft](https://microsoft.github.io/code-with-engineering-playbook/continuous-integration/dev-sec-ops/secret-management/recipes/detect-secrets/), [IBM](https://github.com/IBM/detect-secrets), and [Yelp](https://github.com/Yelp/detect-secrets), is instrumental in identifying sensitive information within project files. Its prowess lies in its extensible Python plugin API, which allows custom rules and heuristics to tackle a broad spectrum of secrets. Designed to scan the project's current state rather than the entire git history, 

Reviewer: My preference would be to remove references to external companies. We are leading on this issue in providing a solution here. Simply stating “leading organizations” would be helpful instead of referencing external private companies. “Detect Secrets is an industry leading tool to identify secure information included in source controlled files.”

Reviewer: I’d recommend removing names from Mermaid graphs, also, that might imply a non-existent endorsement relationship or preference.

Reviewer 2: > My preference would be to remove references to external companies. We are leading on this issue in providing a solution here. Simply stating “leading organizations” would be helpful instead of referencing external private companies. “Detect Secrets is an industry leading tool to identify secure information included in source controlled files.”

Personally - I feel the links to Microsoft and IBM examples are useful for context. @jpl-jengelke is there a wording you can suggest that addresses your concern but keeps these useful links available?

Author: > My preference would be to remove references to external companies. We are leading on this issue in providing a solution here. Simply stating “leading organizations” would be helpful instead of referencing external private companies. “Detect Secrets is an industry leading tool to identify secure information included in source controlled files.”

I just [updated](https://github.com/perryzjc/slim/commit/de962f1162cc30df7277eff670c49fad8bf4c72b) the documentation about this. Thank you! @jpl-jengelke

Author: > > My preference would be to remove references to external companies. We are leading on this issue in providing a solution here. Simply stating “leading organizations” would be helpful instead of referencing external private companies. “Detect Secrets is an industry leading tool to identify secure information included in source controlled files.”
> 
> Personally - I feel the links to Microsoft and IBM examples are useful for context. @jpl-jengelke is there a wording you can suggest that addresses your concern but keeps these useful links available?

I just made [an update](https://github.com/perryzjc/slim/commit/de962f1162cc30df7277eff670c49fad8bf4c72b) to the documentation based on John's suggestion. @jpl-jengelke 
If we have other wording or still want to keep the links, I can update it later. @riverma

Reviewer 2: Looks like @perryzjc made an update to the wording to clarify company citations. LGTM. @jpl-jengelke - we couldn't identify any company references within the Mermaid snippets, so let's leave that for a later discussion / fix and allow @perryzjc to go forth for now.

Reviewer: OK, I'll withdraw my comments for now. We should merge this.

3. Check result in the baseline file
```bash
cat .secrets.baseline
```
> This command is supported in Unix-like system. For Windows, you can use other command, such as `type .secrets.baseline`
If any secrets are detected, the result will be located at the "results:" of the file.

For example,
```json
"results": {
    "Sample/sample-develop.sh": [
    {
    "type": "AWS Sensitive Information",
    "filename": "Sample/sample-develop.sh",
    "hashed_secret": "d3f72b97cd8756fe12345678bdf96aef1ac23b7d",
    "is_verified": false,
    "line_number": 39
},
```
Only line number is visible through this approach. You can check them manually in the file, or use the following audit tool to conveniently view the actual secret.

4. Analyze results by `audit` tool
```bash
detect-secrets audit .secrets.baseline
```
> detect-secrets audit tool can provide you an interactive interface to view the actual secret based on the line number
> and label it as false positive or true positive.

For example,

<img width="564" src="https://github.com/NASA-AMMOS/slim/assets/92573736/0fb25452-1ada-4979-9873-a1ca615701b5">

> It also provides other ways to present the result. For more information, please refer to [Auditing Secrets in Baseline](https://github.com/Yelp/detect-secrets#auditing-secrets-in-baseline)

Reviewer 2: ```suggestion
It also provides other ways to present the result. For more information, please refer to [Auditing Secrets in Baseline](https://github.com/Yelp/detect-secrets#auditing-secrets-in-baseline)

***Note*** if you have marked any detected secrets as true positives, its best to first remove all references to those secrets in your code and then rerun a full scan to generate a fresh audit report that you can compare against to ensure you only have false-positives (white-listed secrets) in your `.secrets.baseline` - this is especially important for Layer 2.
```

```yaml
repos:
  - repo: https://github.com/NASA-AMMOS/slim-detect-secrets
    # using commit id for now, will change to tag when official version is released
    rev: 91e097ad4559ae6ab785c883dc5ed989202c7fbe
    hooks:
      - id: detect-secrets
        args:
          - '--baseline'
          - '.secrets.baseline'
          - '--exclude-files'
          - '.git*'
          - '--exclude-files'
          - '.secrets.*' 

```
This file is used to config the pre-commit hook. In this example, we use the experimental version of `slim-detect-secrets` tool. The `--baseline` argument is used to specify the baseline file generated in the first layer. The `--exclude-files` argument is used to exclude the `.git` directory and `.secrets` baseline files from scanning to reduce false positives. Make modifications when necessary.

3. Install pre-commit hook
```bash
pre-commit install
```
This command reads the `.pre-commit-config.yaml` file and installs the pre-commit hook in your local git repository.
After this, you can see a `.git/hooks/pre-commit` file is created in your local git repository.

> 4. Make sure you have the baseline file `.secrets.baseline` in your local git repository

Reviewer 2: ```suggestion
4. Add the baseline file `.secrets.baseline` in your local Git repository under the following conditions only:
   1. You wish to share white-listed secrets with other team members
   2. You've removed any references in your code to sensitive secrets detected that are not white-listed, i.e. find and modify code referenced in you `.secrets.baseline` with flag `is_secret: true`
   3. You've re-run a fresh `detect-secrets scan` followed by a `detect-secrets audit` and your `.secrets.baseline` ONLY now contains white-listed entries, i.e. `is_secret: false`
   
**Warning**: Adding your `.secrets.baseline` file to your local Git and pushing changes to a remote without following the above directions could create a security risk by sharing the sensitive parts of your security scan with the outside world!
```

5. Commit your changes

Now, you can commit your changes as usual. If any **new secrets** are detected, the commit will be prevented and the secrets will be reported.

For example,

<img width="559" alt="Screen Shot 2023-04-20 at 7 32 10 AM" src="https://user-images.githubusercontent.com/92573736/233398613-6e189322-4d97-47c3-b3ba-bd700a716cf6.png">

> > **Note**: The pre-commit hook blocks a commit by comparing new secrets with the results in the `.secrets.baseline` file. If new secrets are introduced, the hook will report them, but it does not automatically update the `.secrets.baseline` file. To update the baseline file with newly introduced secrets, you need to re-run the scan command in Layer 1 (step 2) and generate a new baseline file.

Reviewer 2: ```suggestion
> **Note**: The pre-commit hook blocks a commit by comparing new secrets with the results in the `.secrets.baseline` file. If new secrets are introduced, the hook will report them, but it does not automatically update the `.secrets.baseline` file. To update the baseline file with newly introduced secrets, you need to re-run the scan command in Layer 1 (step 2) and generate a new baseline file.

> **Note**: during commit checks, detect secrets may only scan for a single secret per file. Thus if you have multiple violations, detect secrets will scan them each time you try to commit. 
```

This workflow will run the `detect-secrets` tool on the GitHub server. If any new secrets are detected, it will:
- Fail the status check

  <img width="507" src="https://github.com/NASA-AMMOS/slim/assets/92573736/bc7670f8-5bbb-414c-b9af-803be05efd67">
- Print a message in the "Details" with instructions on how to resolve the issue

  <img width="1627" src="https://github.com/NASA-AMMOS/slim/assets/92573736/7e884d60-c972-46f2-9785-31b09cebb93f">
- Send an email notification to the user

  <img width="1192" src="https://github.com/NASA-AMMOS/slim/assets/92573736/142e0693-f3f8-42c8-92d7-deb653bdb09c">

2. [Protect your branch](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/managing-a-branch-protection-rule)

<img width="1107" src="https://github.com/NASA-AMMOS/slim/assets/92573736/5a430b2f-7f9b-4ba0-a06a-7c53b300f08b">

After this, if status check fails, GitHub will prevent the merge or push to the protected branch.

<img width="918" src="https://github.com/NASA-AMMOS/slim/assets/92573736/9f8b6239-840a-4ba4-9458-d795f46d7acd">

#### Attention for using Detect Secrets

> > 1.It does not [show all the same type of secrets in a same file to minimize noise](https://github.com/Yelp/detect-secrets/blob/master/docs/design.md#:~:text=Furthermore%2C%20this%20will%20not%20flag%20on%20every%20single%20usage%20of%20a%20given%20secret%20in%20a%20given%20file%2C%20to%20minimize%20noise.)

Reviewer 2: ```suggestion
> 1.It does not [show all the same type of secrets in a same file to minimize noise](https://github.com/Yelp/detect-secrets/blob/master/docs/design.md#potentialsecret)
```

    
    SaveTime["It saves your time. And secrets are safe from GitHub"]
    Clean2 --> SaveTime
    Clean3 --> SaveTime
  end

  User -->|At least use| Layer1
  User -->|Helpful to use| Layer2
  User -->|Optional to use| Layer3

  style User fill:#F6F5F3,stroke:#333,stroke-width:1px
  style UserWorkflow fill:#AF7AC5,stroke:#333,stroke-width:2px
  style Layer1 fill:#F3B044,stroke:#333,stroke-width:2px,stroke-dasharray: 5 5
  style Layer2 fill:#F3B044,stroke:#333,stroke-width:2px,stroke-dasharray: 5 5
  style Layer3 fill:#F3B044,stroke:#333,stroke-width:2px,stroke-dasharray: 5 5
  style Clean1 fill:#5A88ED,stroke:#333,stroke-width:2px
  style Clean2 fill:#5A88ED,stroke:#333,stroke-width:2px
  style Clean3 fill:#5A88ED,stroke:#333,stroke-width:2px
  style SaveTime fill:#5ABF9B,stroke:#333,stroke-width:2px
  style Secure fill:#AF3034,stroke:#333,stroke-width:2px
```
> **Note**: Below three layers, are running on experimental version [slim-detect-secrets](https://github.com/NASA-AMMOS/slim-detect-secrets/tree/exp) which supports additional secret detection [plugins](https://github.com/NASA-AMMOS/slim-detect-secrets/tree/exp#viewing-all-enabled-plugins).
> 
> They are:

> > > * [AWS sensitive information]() (click for more information)

Author: I didn't see the "2" in the type name in my test run. 
<img width="500" alt="image" src="https://github.com/NASA-AMMOS/slim/assets/92573736/2c2e97d6-5d76-4892-85d4-14d95a89394b">

I think "2" comes from the previous version, which you may have run about one month ago.

I feel the solution can be to uninstall the old `detect-secrets` 
```bash
pip uninstall detect-secrets
```
and then install my new version based on the documentation
```bash
pip install git+https://github.com/NASA-AMMOS/slim-detect-secrets.git@exp
```

> > > * [AWS sensitive information]() (click for more information)

Reviewer 2: Thanks @perryzjc - confirmed and resolved by your suggestion.

```

2. Scan all local files from current directory and output the result as a baseline file
```bash
detect-secrets scan ./ --all-files --exclude-files '.secrets.*' --exclude-files '.git*' > .secrets.baseline
```
Here it scans all the local files from current directory, but excludes `.git` directory and `.secrets` baseline files from scanning to reduce false positive. Make modifications when necessary.

3. Check result in the baseline file
```bash
cat .secrets.baseline
```
> This command is supported in Unix-like system. For Windows, you can use other command, such as `type .secrets.baseline`
If any secrets are detected, the result will be located at the "results:" of the file.

For example,
```json
"results": {
    "Jenkins/e2e-opera-pcm-develop.sh": [
    {
    "type": "AWS Sensitive Information",
    "filename": "Jenkins/e2e-opera-pcm-develop.sh",
    "hashed_secret": "a2c61f79dd90429e5284ab45de3318c16756a2f0",
    "is_verified": false,
    "line_number": 39
},
```
Only line number is visible through this approach. You can check them manually in the file, or use the following audit tool to conveniently view the actual secret.

> 4. Analyze results by `audit` tool

Author: > In addition to your step 4 here, I think you need add some clarification to the user that they never need to run the initial scan again after the audit if they set up Layer 2.
> 
> i.e. no need to run
> 
> ```
> detect-secrets scan ./ --all-files --exclude-files '.secrets.*' --exclude-files '.git*' > .secrets.baseline
> ```

@riverma I am not quite sure what it mean here. Let me clarify how layer 2 works:

1. When Layer 2 gets triggered, it will look for new secrets by comparing the existing baseline file. That's the reason it has to have a baseline file existing locally.

2. Layer 2 does not change the baseline file automatically. For example, if we have an initial baseline file with empty results, then we add a file with secrets and commit it to trigger Layer 2, then Layer 2 can tell us that secret got detected, but the baseline file remains the same. This is the reason we will use Layer 1 to update the baseline file, which was mentioned [here in the document](https://github.com/NASA-AMMOS/slim/blob/d20ee6134a0dc0e0dab11d2d2570e358ef7e4550/continuous-testing/starter-kits/README.md#:~:text=If%20you%20want%20to%20add%20new%20secret%20results%2C%20you%20need%20to%20update%20.secrets.baseline%20file%20by%20re%2Drunning%20the%20scan%20command%20and%20generate%20a%20new%20baseline%20file.)

I will update the document with this fact more explicitly.

Please let me know if you have any other questions. Thank yoU!

> 4. Analyze results by `audit` tool

Reviewer 2: @perryzjc LGTM! Thanks for the clarification here.

```yaml
repos:
  - repo: https://github.com/NASA-AMMOS/slim-detect-secrets
    # using commit id for now, will change to tag when official version is released
    rev: 91e097ad4559ae6ab785c883dc5ed989202c7fbe
    hooks:
      - id: detect-secrets
        args:
          - '--baseline'
          - '.secrets.baseline'
          - '--exclude-files'
          - '.git*'
          - '--exclude-files'
          - '.secrets.*' 

```
This file is used to config the pre-commit hook. In this example, we use the experimental version of `slim-detect-secrets` tool. The `--baseline` argument is used to specify the baseline file generated in the first layer. The `--exclude-files` argument is used to exclude the `.git` directory and `.secrets` baseline files from scanning to reduce false positives. Make modifications when necessary.

3. Install pre-commit hook
```bash
pre-commit install
```
This command reads the `.pre-commit-config.yaml` file and installs the pre-commit hook in your local git repository.
After this, you can see a `.git/hooks/pre-commit` file is created in your local git repository.

4. Make sure you have the baseline file `.secrets.baseline` in your local git repository

> 5. Commit your changes

Reviewer 2: This issue has been resolved. Thanks @perryzjc. One note is that it seems multiple plugins don't run against the same file during Layer 2 commit checks. I've added suggested wording to clarify this as a PR comment.

```mermaid
sequenceDiagram
    participant User as Developer
    participant GH as GitHub
    participant Workflow as detect-secrets.yaml
    participant GA as GitHub Action
    participant DS as Detect-Secrets

    Note over User,GH: Developer creates pull request or pushes to branch
    User->>+GH: Creates pull request / pushes to branch
    GH->>+Workflow: Triggers GitHub Action workflow
    Workflow->>GA: Sets up and runs Detect-Secrets scan
    GA->>DS: Requests secret scan
    DS->>DS: Scans repository for secrets
    alt Secrets Detected
        DS-->>GA: Returns detected secrets
        GA-->>GH: Fails status check
        GH-->>User: Prevents merge / push & sends email notification
    else No Secrets Detected
        DS-->>GA: Returns clean result
        GA-->>GH: Passes status check
        GH-->>User: Allows merge / push
    end

```
Starter Kit:

> 1. Create a workflow file `detect-secrets.yaml` in `.github/workflows` directory from your repository root.

Reviewer 2: After running this on a fresh repository on GitHub, I received the following in the GH Actions workflow log (see trace [here](https://github.com/riverma/test-repo/actions/runs/5139169327/jobs/9249315949)):

```
Run # scripts to scan repository for new secrets
cp: cannot stat '.secrets.baseline': No such file or directory
Error: Process completed with exit code 1.
```

Is there a way you can check for and create a blank `.secrets.baseline` file if being run for the first time on a repo that has never used detect secrets before?

Reviewer: try `touch .secrets.baseline` which will ensure it's always there.

Author: > try `touch .secrets.baseline && <additional_command>` which will ensure it's always there.

Yep, I have a similar idea. But due to the format requirement of the baseline file (detect-secrets version can be different), my new solution is to generate the blank baseline file (but with the correct config) by scanning the result on an empty folder.
```yaml
      - name: Create a blank .secrets.baseline (but with correct configuration) if it does not exist
        run: |
          if [ ! -f .secrets.baseline ]; then
            # The generated baseline file will only be available temporarily on the GitHub side and will not appear on the user's local files
            # Scanning an empty folder ensures no initial secrets are included
            echo "⚠️ No existing .secrets.baseline file detected. Creating a new blank baseline file."
            mkdir empty-dir
            detect-secrets scan empty-dir > .secrets.baseline
            echo "✅ Blank .secrets.baseline file created successfully."
            rm -r empty-dir
          else
            echo "✅ Existing .secrets.baseline file detected. No new baseline file will be created."
          fi
```

Reviewer: Touch does exactly that. If it's not there, it creates an empty file. 

You may want to use `>>` so if a file exists it doesn't replace its contents with an empty file.

Author: > You may want to use `>>` so if a file exists it doesn't replace its contents with an empty file.

I have a clarification. 

To run detect-secrets successfully, the baseline file should not be entirely empty but should contain some information, even if no secrets are detected. Here's an example of the required format for the baseline file:

<img width="488" alt="image" src="https://github.com/NASA-AMMOS/slim/assets/92573736/2d87794c-10e3-40aa-ba5b-52bbc9bde670">

And, this format of the content generated can vary based on the version of detect-secrets. So I use this command to help with the compatibility
`detect-secrets scan empty-dir > .secrets.baseline`

My solution uses `if-else` approach to determine whether we need to create the "initial" baseline file. And even though the file got created by the GitHub Action server, the change is not committed, so it will not affect out project code.

Reviewer 2: Hi @perryzjc - thanks for getting the GitHub Action going. Working now in my [test run](https://github.com/riverma/test-repo/actions/runs/5224805784/jobs/9433474806)! Though please see two suggestions below: 

Suggestions:
1. The sample workflow file mentioned in your Usage section does not work as-is for me. I received an invalid workflow file exception (see [here](https://github.com/riverma/test-repo/actions/runs/5224837125) and [here](https://github.com/riverma/test-repo/actions/runs/5224849234)). I had to remove the "description" and the "branding" parameters to get it to work. Can you resolve this?
2. I'm not clear on the "Use latest version" button of your [GH action page](https://github.com/marketplace/actions/detect-secrets-action) and how your action will support inline use within the "steps" section of a GitHub workflow file. Seems like the way you have it configured, your GitHub action can only work as a single workflow file, and not within a workflow file that contains multiple steps. Might be good to clarify that in the description  of your GitHub action.

Reviewer 2: After chatting with @perryzjc - let's go forth with reverting to the previous GitHub Action strategy, namely: embedding the Workflow file contents inline into this documentation for users to copy/paster. The GH marketplace strategy is complicated and will require more investigation.

Author: @riverma I just made some [updates](https://github.com/NASA-AMMOS/slim/pull/95/commits/d27620858dfaa0a34b6a04273259ca2e96007753) based on the conversation.

> 1. Create a workflow file `detect-secrets.yaml` in `.github/workflows` directory from your repository root.

Reviewer 2: One more note: try pushing this workflow to https://github.com/marketplace?type=actions when you are ready. This will make installation a breeze.

Author: When I publish it to the Marketplace, who should be the owner of this Action?
When I chose NASA-AMMOS, it asked me to contact the admin of the organization to accept the Marketplace agreement:
<img width="1165" alt="image" src="https://github.com/NASA-AMMOS/slim/assets/92573736/1c12cb15-5b5f-4e14-8287-2005e2747330">

If I publish it by own account, I can accept the agreement anyway.

Author: Currently, I published my GitHub Action with additional documentation through my account. 
Here is the [link](https://github.com/marketplace/actions/detect-secrets-action) to it.

Reviewer 2: @perryzjc - its fine to keep on your account for now, until we come up with a better organization contact info.
